autogpt_platform/backend/backend/server/v2/library/db.py

```diff
@@ -543,20 +545,105 @@ async def update_library_agent(
+
+        for schedule in schedules:
+            try:
+                await scheduler_client.delete_schedule(
+                    schedule_id=schedule.id, user_id=user_id
+                )
+                logger.info(f"Deleted schedule {schedule.id} for graph {graph_id}")
+            except Exception as e:
+                logger.error(
+                    f"Failed to delete schedule {schedule.id} for graph {graph_id}: {e}"
+                )
+
+    except Exception as e:
+        logger.error(f"Failed to cleanup schedules for graph {graph_id}: {e}")
+
+
+async def _cleanup_webhooks_for_graph(graph_id: str, user_id: str) -> None:
+    """
+    Clean up all webhooks for a specific graph and user.
+
+    Args:
+        graph_id: The ID of the graph
+        user_id: The ID of the user
+    """
+    try:
+        # Find all webhooks that trigger nodes in this graph
+        webhooks = await integrations_db.find_webhooks_by_graph_id(
+            graph_id=graph_id, user_id=user_id
+        )
+
+        for webhook in webhooks:
+            try:
+                await integrations_db.delete_webhook(
+                    user_id=user_id, webhook_id=webhook.id
+                )
+                logger.info(f"Deleted webhook {webhook.id} for graph {graph_id}")
+            except Exception as e:
+                logger.error(
+                    f"Failed to delete webhook {webhook.id} for graph {graph_id}: {e}"
+                )
+
+    except Exception as e:
+        logger.error(f"Failed to cleanup webhooks for graph {graph_id}: {e}")
+
+
 async def delete_library_agent(
     library_agent_id: str, user_id: str, soft_delete: bool = True
 ) -> None:
-    if soft_delete:
-        deleted_count = await prisma.models.LibraryAgent.prisma().update_many(
-            where={"id": library_agent_id, "userId": user_id}, data={"isDeleted": True}
-        )
-    else:
-        deleted_count = await prisma.models.LibraryAgent.prisma().delete_many(
-            where={"id": library_agent_id, "userId": user_id}
-        )
-    if deleted_count < 1:
+    from backend.data.db import transaction
+
+    # First get the agent to find the graph_id for cleanup
+    library_agent = await prisma.models.LibraryAgent.prisma().find_unique(
+        where={"id": library_agent_id}, include={"AgentGraph": True}
+    )
+
+    if not library_agent or library_agent.userId != user_id:
         raise NotFoundError(f"Library agent #{library_agent_id} not found")
 
+    graph_id = library_agent.agentGraphId
+
+    # Use transaction to prevent race condition between deletion and cleanup
+    async with transaction():
+        # Clean up associated schedules and webhooks BEFORE deleting the agent
+        # This prevents race conditions where executions could start after agent deletion
+        # but before cleanup completes
+        await _cleanup_schedules_for_graph(graph_id=graph_id, user_id=user_id)
+        await _cleanup_webhooks_for_graph(graph_id=graph_id, user_id=user_id)
```

Comment:
~~I think these should be version-specific~~

If a user has multiple LibraryAgent entries for different versions of the same graph (which is technically possible), we don't want to mess with stuff set up by the *other* LibraryAgent, right?